### PR TITLE
Support for Google Vertex Models

### PR DESCRIPTION
**Pull request by bilbo7833**
## Support for Google Vertex Models - Gemini 1.5

This PR adds a provider module for Google Vertex AI Models and implements support for Gemini 1.5-flash and Gemini 1.5-pro. In order to use these model providers, you need Google Cloud Authentication, which you can get by installing a command line tool and then authenticating with [Application Default Credentials](https://cloud.google.com/docs/authentication/provide-credentials-adc). The configuration file to use when running the `DataAgentServer` needs to contain two extra fields, for project id and project location.

```{
  "prompt": "You are a helpful customer service representative for a credit card company who helps answer customer questions about their past transactions and spending history. Today's date is January 18th, 2024. You provide precise answers and use functions to look up information. You DO NOT provide general answers and only give an answer after you retrieved all the data you need via functions. Only invoke one function at a time and wait for the results before invoking another function.When you have worked out your answer, you answer the user question in one of two ways: 1) in markdown syntax using tables where appropriate to show data or 2) by calling the data_visualization function to display the data in a suitable fashion. Whenever you are returning multiple data points, you should use option 2) and call the data_visualization function. You should only call the data_visualization function when you already have all the data to display.",
  "api-url": "http://localhost:8888/graphql",
  "auth-field": "customerid",
  "auth-is-integer": true,
  "plot-function": "oneD",
  "model" : {
    "provider" : "vertex",
    "name": "gemini-1.5-flash",
    "project_id": "vertex-gemini-424313",
    "location": "europe-west4"
  }
}
```
### Implementation Details
- Since Vertex treats `FunctionCall` and `FunctionResponse` as separate models, it was necessary to be able to save and load these two different objects from the Backend. Thus, the functionality of `GenericChatMessage`, and the GraphQL Queries in the tools and the sqrl files needed to be extended. This will be supported in a different PR, which should be merged before this one.
- The implementation of the vertex module was done around the `Content` class from Google, which represents the basic type of message, containing `Parts`, which in turn contain, text, images, function calls or function responses.

### CODE REVIEW

**Review by mbroecheler — APPROVED** [review]
This looks good to me - it follows the same architectural pattern while navigating around the Google quirks (it's funny that they insist on protobuffers when the rest of the world has embraced json and the capitalizeType is pretty hilarious). I didn't parse the code explicitly and assume that it works.

**Comment by bilbo7833** [comment]
@mbroecheler I have added configuration files for the other use cases, but was unable to test them due to my docker issues 😢 I really need an EC2 instance pls 🙏  After merging this PR, we need to update the .sqrl and the .grphqls files in the datasqrl-examples Repo as well.

**Comment by bilbo7833** [comment]
This branch is now rebased on top of https://github.com/DataSQRL/apiRAG/tree/feat/extendedapi and it should be merged afterwards

**Comment by bilbo7833** [comment]
This branch is ready to be merged

File: api-examples/weather/weather.config.json
Comment (bilbo7833) [reply api-examples/weather/weather.config.json]: I just renamed it to be more explicit. It is now called `weather.bedrock-llama370b.config.json`

**Comment by bilbo7833** [comment]
@mbroecheler This PR has been rebased and I have also fixed a small bug with the TokenCounter based on HuggingFace. This is ready to be merged, pending your approval.

**Pull request merged**
